cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+        __temp,
+        sizeof(_Tp) * __count,
+        __kind,
+        __buf_.get_stream().get());
+      _CUDA_VSTD::return_temporary_buffer(__temp);
+    }
+    else
+    {
+      (void) __last;
+      _CCCL_TRY_CUDA_API(
+        ::cudaMemcpyAsync,
+        "cudax::async_mdarray::__copy_cross: failed to copy data",
+        __dest,
+        _CUDA_VSTD::to_address(__first),
+        sizeof(_Tp) * __count,
+        __kind,
+        __buf_.get_stream().get());
+    }
+  }
+
+  //! @brief Copy-constructs elements in the range `[__first, __first + __count)`.
+  //! @param __first Pointer to the first element to be initialized.
+  //! @param __count The number of elements to be initialized.
+  _CCCL_HIDE_FROM_ABI void __fill_n(pointer __first, size_type __count, const _Tp& __value)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::fill_n(__first, __count, __value);
+    }
+    else
+    {
+      thrust::fill_n(thrust::cuda::par_nosync.on(__buf_.get_stream().get()), __first, __count, __value);
+    }
+  }
+
+  //! @brief Equality-compares elements from the sequence `[__first1, __last1)` with those of sequence
+  //! `[__first2, __last2)`.
+  //! @param __first1 Pointer to the start of the first sequence.
+  //! @param __last1 Pointer to the end of the first sequence.
+  //! @param __first2 Pointer to the start of the second sequence.
+  //! @param __last2 Pointer to the end of the second sequence.
+  _CCCL_HIDE_FROM_ABI bool
+  __equality(const_pointer __first1, const_pointer __last1, const_pointer __first2, const_pointer __last2) const
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::equal(__first1, __last1, __first2, __last2);
+    }
+    else
+    {
+      return ((__last1 - __first1) == (__last2 - __first2))
+          && thrust::equal(thrust::cuda::par_nosync.on(__buf_.get_stream().get()), __first1, __last1, __first2);
+    }
+    _CCCL_UNREACHABLE();
+  }
+
+  template <class _IndexType, size_t... _Exts>
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI static constexpr size_type
+  __size(const _CUDA_VSTD::extents<_IndexType, _Exts...>& __ext) noexcept
+  {
+    size_type __res = 1;
+    for (size_t __i = 0; __i != sizeof...(_Exts); ++__i)
+    {
+      __res *= __ext.extent(__i);
+    }
+    return __res;
+  }
+
+public:
+  //! @addtogroup construction
+  //! @{
+
+  //! @brief Copy-constructs a async_mdarray
+  //! @param __other The other async_mdarray.
+  //! The new async_mdarray has capacity of \p __other.size() which is potentially less than \p __other.capacity().
+  //! @note No memory is allocated if \p __other is empty
+  _CCCL_HIDE_FROM_ABI async_mdarray(const async_mdarray& __other)
+      : __buf_(__other.get_memory_resource(), __other.get_stream(), __other.size())
+      , __mapping_(__other.__mapping_)
+      , __policy_(__other.__policy_)
+  {
+    if (__other.size() != 0)
+    {
+      this->__copy_same(__other.__unwrapped_begin(), __other.__unwrapped_end(), __unwrapped_begin());
```

Comment:
If the layout is not exhaustive, then this will copy elements that do not belong to the input.  If the layout is not unique, then this may copy elements multiple times.  While creating an `mdarray` with a nonexhaustive layout is a bit weird, nothing here prevents users from doing that.

In general, `mdarray` (as defined in P1684) cannot be implemented without an mdspan copy algorithm.